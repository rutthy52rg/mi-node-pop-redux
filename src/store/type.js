export const HISTORY_BACK = "HISTORY_BACK";
export const AUTH_LOGIN_SUCCESS = "AUTH_LOGIN_SUCCESS";
export const AUTH_LOGIN_FAILURE = "AUTH_LOGIN_FAILURE";
export const AUTH_LOGIN_REQUEST = "AUTH_LOGIN_REQUEST";
export const AUTH_LOGOUT = "AUTH_LOGOUT";

export const ADVERTS_LOADED_SUCCESS = "ADVERTS_LOADED_SUCCESS";
export const ADVERTS_LOADED_FAILURE = "ADVERTS_LOADED_FAILURE";
export const ADVERTS_LOADED_REQUEST = "ADVERTS_LOADED_REQUEST";
export const UI_RESET_ERROR = "UI_RESET_ERROR";

export const ADVERT_LOADED_REQUEST = "ADVERT_LOADED_REQUEST";
export const ADVERT_LOADED_SUCCESS = "ADVERT_LOADED_SUCCESS";
export const ADVERT_LOADED_FAILURE = "ADVERT_LOADED_FAILURE";

export const ADVERT_DELETED_REQUEST = "ADVERT_DELETED_REQUEST";
export const ADVERT_DELETED_SUCCESS = "ADVERT_DELETED_SUCCESS";
export const ADVERT_DELETED_FAILURE = "ADVERT_DELETED_FAILURE";

export const ADVERT_CREATED_REQUEST = "ADVERT_CREATED_REQUEST";
export const ADVERT_CREATED_SUCCESS = "ADVERT_CREATED_SUCCESS";
export const ADVERT_CREATED_FAILURE = "ADVERT_CREATED_FAILURE";

export const TAGS_LOADED_REQUEST = "TAGS_LOADED_REQUEST";
export const TAGS_LOADED_SUCCESS = "TAGS_LOADED_SUCCESS";
export const TAGS_LOADED_FAILURE = "TAGS_LOADED_FAILURE";
