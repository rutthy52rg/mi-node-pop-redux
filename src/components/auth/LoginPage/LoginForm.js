import T from 'prop-types';

import useForm from '../../../hooks/useForm';

const validEmail = ({ email }) => email;
const validPassword = ({ password }) => password;

function LoginForm({ onSubmit, isLoading }) {
  const {
    formValue: credentials,
    handleChange,
    handleSubmit,
    validate,
  } = useForm({
    email: '',
    password: '',
    remember: false,
  });
  const { email, password, remember } = credentials;

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <input name="email" value={email} onChange={handleChange} />
      <input
        type="password"
        name="password"
        value={password}
        onChange={handleChange}
      />
      <input
        type="checkbox"
        name="remember"
        checked={remember}
        onChange={handleChange}
      />
      <button disabled={!validate(validEmail, validPassword, () => !isLoading)}>
        Login
      </button>
    </form>
  );
}

LoginForm.propTypes = {
  isLoading: T.bool,
  onSubmit: T.func.isRequired,
};

export default LoginForm;
