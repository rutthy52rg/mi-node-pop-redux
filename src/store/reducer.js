import {
  ADVERTS_LOADED_SUCCESS,
  ADVERT_LOADED_SUCCESS,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
} from "./type";

export const stateDefault = {
  auth: false,
  adverts: {
    areLoaded: false,
    data: [],
  },
  ui: {
    isLoadding: false,
    error: null,
  },
};

export function auth(state = stateDefault.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adverts(state = stateDefault.adverts, action) {
  if (action.type === ADVERTS_LOADED_SUCCESS) {
    return { areLoaded: true, data: action.payload };
  }
  if (action.type === ADVERT_LOADED_SUCCESS) {
    return { ...state, data: [action.payload] };
  }
  return state;
}

export function ui(state = stateDefault.ui, action) {
  if (/_FAILURE$/.test(action.type)) {
    return {
      isLoadding: false,
      error: action.payload,
    };
  }
  if (/_REQUEST$/.test(action.type)) {
    return {
      isLoadding: true,
      error: null,
    };
  }
  if (/_SUCCESS$/.test(action.type)) {
    return {
      isLoadding: false,
      error: null,
    };
  }
  if (action.type === UI_RESET_ERROR) {
    return {
      ...state,
      error: null,
    };
  }
  return state;
}
