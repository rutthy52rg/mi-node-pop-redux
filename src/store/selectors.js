export const getIsLogged = (state) => state.auth;
export const getAdverts = (state) => state.adverts.data;
export const areAdvertsLoaded = (state) => state.adverts.areLoaded;
export const areTagsLoaded = (state) => state.tags.areLoaded;

export const getAdvertDetail = (id) => (state) =>
  state.adverts.data.find((advert) => advert.id.toString() === id);
export const getUi = (state) => state.ui;
