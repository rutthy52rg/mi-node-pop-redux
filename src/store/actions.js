import { areAdvertsLoaded, getAdvertDetail } from "./selectors";
import {
  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_LOADED_FAILURE,
  ADVERT_LOADED_REQUEST,
  ADVERT_LOADED_SUCCESS,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
} from "./type";

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});
export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});
export const authLoginFailure = (error) => ({
  type: AUTH_LOGIN_FAILURE,
  payload: error,
  error: true,
});
//gracias al middleware y thunk
export const authLogin = (credentials) => {
  return async function (dispatch, getState, { api, router }) {
    try {
      dispatch(authLoginRequest());
      // await login(credentials);
      await api.auth.login(credentials);
      dispatch(authLoginSuccess());
      const to = router.state.location.state?.from?.pathname || "/";
      router.navigate(to, { replace: true });
    } catch (error) {
      dispatch(authLoginFailure(error));
    }
  };
};

export const authLogoutSuccess = () => ({
  type: AUTH_LOGOUT,
});
export const authLogout = () => {
  return async function (dispatch, getState, { api, router }) {
    await api.auth.logout();
    dispatch(authLogoutSuccess());
    router.navigate(`/adverts`);
  };
};

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});
export const advertsLoadedSuccess = (adverts) => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});
export const advertsLoadedFailure = (error) => ({
  type: ADVERTS_LOADED_FAILURE,
  payload: error,
  error: true,
});

// thuk

export const advertsLoad = () => {
  return async function (dispatch, getState, { api }) {
    const areLoaded = areAdvertsLoaded(getState());
    if (areLoaded) return;
    try {
      dispatch(advertsLoadedRequest());
      const adverts = await api.adverts.getAdverts();
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
    }
  };
};

export const advertLoadedRequest = () => ({
  type: ADVERT_LOADED_REQUEST,
});
export const advertLoadedSuccess = (tweet) => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: tweet,
});
export const advertLoadedFailure = (error) => ({
  type: ADVERT_LOADED_FAILURE,
  payload: error,
  error: true,
});

// thuk
export const advertLoad = (advertId) => {
  return async function (dispatch, getState, { api, router }) {
    const isLoaded = getAdvertDetail(advertId)(getState());
    if (isLoaded) return;
    try {
      dispatch(advertLoadedRequest());
      const advert = await api.adverts.getAdvert(advertId);
      dispatch(advertLoadedSuccess(advert));
      console.log("hecho");
    } catch (error) {
      dispatch(advertLoadedFailure(error));
    }
  };
};
export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
