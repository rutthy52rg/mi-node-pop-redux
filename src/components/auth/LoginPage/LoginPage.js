import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { authLogin, uiResetError } from "../../../store/actions";
import { getUi } from "../../../store/selectors";
import LoginForm from "./LoginForm";

function LoginPage({ ...props }) {
  const { isLoadding, error } = useSelector(getUi);
  const dispatch = useDispatch();
  const handleSubmit = (credentials) => {
    dispatch(authLogin(credentials));
  };
  const resetErrorHandle = () => {
    dispatch(uiResetError());
  };

  return (
    <div>
      <LoginForm onSubmit={handleSubmit} isLoading={isLoadding} />
      {isLoadding && <p>...login in nodepop</p>}
      {error && (
        <div onClick={resetErrorHandle} style={{ color: "red" }}>
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
