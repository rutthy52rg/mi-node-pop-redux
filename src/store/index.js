import { composeWithDevTools } from "@redux-devtools/extension";
import { combineReducers, createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import * as adverts from "../components/adverts/service";
import * as auth from "../components/auth/service";
import * as reducers from "./../store/reducer";

const reducer = combineReducers(reducers);

const logger = (store) => (next) => (action) => {
  console.group(action.type);
  console.info("dispatching", action, store.getState());
  const result = next(action);
  console.log("next state", store.getState());
  console.groupEnd();
  return result;
};
const failureRedirections =
  (router, redirections) => (store) => (next) => (action) => {
    const result = next(action);
    if (action.error) {
      const redirection = redirections[action.payload.status];
      if (redirection) {
        router.navigate(redirection);
      }
    }
    return result;
  };
export default function configureStore(preloadedState, { router }) {
  const middlewares = [
    thunk.withExtraArgument({ api: { auth, adverts }, router }),
    failureRedirections(router, {
      401: "/login",
      404: "/404",
    }),
    logger
  ];
  const store = createStore(
    reducer,
    preloadedState,
    composeWithDevTools(applyMiddleware(...middlewares))
  );
  return store;
}
