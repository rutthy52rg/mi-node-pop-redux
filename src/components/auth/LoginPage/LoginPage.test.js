import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { stateDefault } from "../../../store/reducer";
import LoginPage from "./LoginPage";

jest.mock("../../../store/actions");

//snapshot
describe("loginPage", () => {
  const store = {
    getState: () => stateDefault,
    dispatch: () => {},
    subscribe: () => {},
  };
  const renderComponent = () =>
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );

  test("snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
