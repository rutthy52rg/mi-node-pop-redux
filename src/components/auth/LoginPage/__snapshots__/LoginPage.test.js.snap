// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginPage snapshot 1`] = `
<div>
  <div>
    <form>
      <input
        name="email"
        value=""
      />
      <input
        name="password"
        type="password"
        value=""
      />
      <input
        name="remember"
        type="checkbox"
      />
      <button
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
